### PR TITLE
Add the GMTSampleData class to simplify the load_sample_data and list_sample_data functions (#2342)Co-authored-by: Will Schlitzer <schlitzer90@gmail.com> Co-authored-by: Yvonne Fröhlich <94163266+yvonnefroehlich@users.noreply.github.com> Co-authored-by: Michael Grund <23025878+michaelgrund@users.noreply.github.com>

### LINKED ISSUE #1774: Improve the documentation of load_sample_data and list_sample_data
We have the `list_sample_data` function to list all sample data that are available in `load_sample_data`. However, the documentation of [list_sample_data](https://www.pygmt.org/dev/api/generated/pygmt.datasets.list_sample_data.html#pygmt.datasets.list_sample_data) and [`load_sample_data`](https://www.pygmt.org/dev/api/generated/pygmt.datasets.load_sample_data.html#pygmt.datasets.load_sample_data) say nothing, so users have to call the `list_sample_data` function to see it. 

It would be better and easier to read if we can also list all available data in the documentation.

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

It's easier to review this PR by looking at the changes in each commit:

1. 22a8e48bbd75daaf4a9082ad79d83774c4ee6a13: Move `load_sample_data` and `list_sample_data` to the end. Required by the changes in commit 4641649b5af988fcc08ed8dd128c511bdab61eee
2. 4641649b5af988fcc08ed8dd128c511bdab61eee: Add a `GMTSampleData` class, similar to the `GMTRemoteData` used in `load_remote_dataset.py`. This class contains two attributes, the function to load the dataset and the description of the dataset. With the `GMTSampleData` class, the `load_sample_data` and `list_sample_data` functions are greatly simplified and we no longer need to maintain two dictionaries which have the same keys.
3. ~~ea9132c9008a9e22c7006843f02697d7b33c8413: Add a `fmt_dataset_list` decorator to automatically inserts the list of available datasets to the docstrings of the `load_sample_data` function, to address #1774.~~ 

**Preview:** https://pygmt-dev--2342.org.readthedocs.build/en/2342/api/generated/pygmt.datasets.load_sample_data.html

image

Fixes #1774 and supersedes #1814.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/datasets/samples.py
Comment (seisman) [inline pygmt/datasets/samples.py:277]:
```suggestion
    "rock_compositions": GMTSampleData(
```

File: pygmt/datasets/samples.py
Comment (seisman) [inline pygmt/datasets/samples.py:308]: This decorator inserts the list of available datasets to the `load_sample_data` function by substituting the `{dataset_list}` placeholder with the following text strings:
```
- ``"bathymetry"``: Table of ship bathymetric observations off Baja California
- ``"earth_relief_holes"``: Regional 20 arc-minutes Earth relief grid with holes
- ...
```

However, because the placeholder `{dataset_list}` is indented by 8 whitespaces (see Line 341), substituting the placeholder with the text strings would give:
```
        - ``"bathymetry"``: Table of ship bathymetric observations off Baja California
- ``"earth_relief_holes"``: Regional 20 arc-minutes Earth relief grid with holes
- ...
```
So at line 302, I have to use `"\n        "`.

File: pygmt/datasets/samples.py
Comment (willschlitzer) [inline pygmt/datasets/samples.py:302]:
```suggestion
    """
    # Add a new lines and 8 white spaces for formatting in the documentation
    texts = "\n        ".join(
```

I don't know of a better way to fix this problem for documentation displays, and I think using `\n        ` works since this runs behind the scenes.  I do think there should be a comment explaining why there is a strange join string.

File: pygmt/datasets/samples.py
Comment (willschlitzer) [inline pygmt/datasets/samples.py:324]: Should there be a `test_list_sample_data` that has a dictionary of all of the functions/descriptions to test line 324?

Comment (seisman) [reply pygmt/datasets/samples.py:325]: Are you suggesting a test to check if the returned value is equal to:
```
{'bathymetry': 'Table of ship bathymetric observations off Baja California',
 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes',
 'fractures': 'Table of hypothetical fracture lengths and azimuths',
 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from  Mueller et al., 1993',
 'japan_quakes': 'Table of earthquakes around Japan from NOAA NGDC database',
 'mars_shape': 'Table of topographic signature of the hemispheric dichotomy of  Mars from Smith and Zuber (1996)',
 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis in Geology',
 'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
 'rock_compositions': 'Table of rock sample compositions',
 'usgs_quakes': 'Table of global earthquakes from the USGS'}
```

Comment (seisman) [reply pygmt/datasets/samples.py:325]: In 518900d6d9de9034345f6da744cd45d066e00493, I added a test to check if the returned value is in `dict` type.

Comment (willschlitzer) [reply pygmt/datasets/samples.py:325]: I think that's fine; I was just going off of the codecov alert saying that like 324 was untested. My thought had been to use something like a list of the dataset names being equal to `dict.keys()` to avoid a very lengthy comparison, but I think your way works better.

**Review by willschlitzer — APPROVED** [review]

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers It would be good if we can have one more reviewer since this PR brings some big changes.

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers It would be good if we can have one more reviewer since this PR brings some big changes.

File: pygmt/datasets/samples.py
Comment (yvonnefroehlich) [inline pygmt/datasets/samples.py:254]: To have a consistent citation style.
```suggestion
    "hotspots": GMTSampleData(
        func=_load_hotspots,
        description="Table of locations, names, and symbol sizes of hotpots from "
        "Mueller et al. (1993)",
```

File: pygmt/datasets/samples.py
Comment (yvonnefroehlich) [inline pygmt/datasets/samples.py:259]:
```suggestion
    "japan_quakes": GMTSampleData(
        func=_load_japan_quakes,
        description="Table of earthquakes around Japan from the NOAA NGDC database",
    ),
```

File: pygmt/datasets/samples.py
Comment (yvonnefroehlich) [inline pygmt/datasets/samples.py:302]:
```suggestion
    # Add a new line and 8 white spaces for formatting in the documentation
```

**Review by maxrjones — COMMENTED** [review]
Thanks for working on this, it'll be a big improvement!

My preference would be to document of the list of datasets manually as the output of a docstring example for `list_sample_data` (in which case the doctest should catch whether it's out of date). I would prefer avoiding the decorator/docstring inject trick when possible because it makes the docstrings less interpretable for IDEs like VSCode.

**Comment by seisman** [comment]
> Thanks for working on this, it'll be a big improvement!
> 
> My preference would be to document of the list of datasets manually as the output of a docstring example for `list_sample_data` (in which case the doctest should catch whether it's out of date). I would prefer avoiding the decorator/docstring inject trick when possible because it makes the docstrings less interpretable for IDEs like VSCode.

That's a good point.

> My preference would be to document of the list of datasets manually as the output of a docstring example for `list_sample_data` (in which case the doctest should catch whether it's out of date).

I'm OK with adding the the output of `list_sample_data` as a docstring example, but instead of adding it to the `list_sample_data` function, I feel it's more useful to add it to the `load_sample_data` function, i.e., in the `load_sample_data` function:

```
>>> # call list_sample_data to see the available datasets
>>> list_sample_data()
{'bathymetry': 'Table of ship bathymetric observations off Baja California',
 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes',
 'fractures': 'Table of hypothetical fracture lengths and azimuths',
 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from  Mueller et al., 1993',
 'japan_quakes': 'Table of earthquakes around Japan from NOAA NGDC database',
 'mars_shape': 'Table of topographic signature of the hemispheric dichotomy of  Mars from Smith and Zuber (1996)',
 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis in Geology',
 'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
 'rock_compositions': 'Table of rock sample compositions',
 'usgs_quakes': 'Table of global earthquakes from the USGS'}
>>> # load the sample bathymetry dataset
>>> data = load_sample_data("bathymetry")
```

**Comment by maxrjones** [comment]
> > I would prefer avoiding the decorator/docstring inject trick when possible because it makes the docstrings less interpretable for IDEs like VSCode.
> 
> That's a good point.
> 
> > My preference would be to document of the list of datasets manually as the output of a docstring example for `list_sample_data` (in which case the doctest should catch whether it's out of date).
> 
> I'm OK with adding the the output of `list_sample_data` as a docstring example, but instead of adding it to the `list_sample_data` function, I feel it's more useful to add it to the `load_sample_data` function, i.e., in the `load_sample_data` function:
> 
> ```python
> >>> # use list_sample_data to see the available datasets
> >>> list_sample_data()
> {'bathymetry': 'Table of ship bathymetric observations off Baja California',
>  'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes',
>  'fractures': 'Table of hypothetical fracture lengths and azimuths',
>  'hotspots': 'Table of locations, names, and symbol sizes of hotpots from  Mueller et al., 1993',
>  'japan_quakes': 'Table of earthquakes around Japan from NOAA NGDC database',
>  'mars_shape': 'Table of topographic signature of the hemispheric dichotomy of  Mars from Smith and Zuber (1996)',
>  'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
>  'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis in Geology',
>  'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
>  'rock_compositions': 'Table of rock sample compositions',
>  'usgs_quakes': 'Table of global earthquakes from the USGS'}
> >>> # load the sample bathymetry dataset
> >>> data = load_sample_data("bathymetry")
> ```

sounds good to me

**Review by michaelgrund — APPROVED** [review]

**Comment by seisman** [comment]
I tried to follow @maxrjones's suggestion and added a doctest for `list_sample_data()`, but things is becoming more complicated.

The default returned value of `list_sample_data` is a unformatted big dictionary, which is not readable and needs to be in a single line:
```
{'bathymetry': 'Table of ship bathymetric observations off Baja California', 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes', 'fractures': 'Table of hypothetical fracture lengths and azimuths', 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from  Mueller et al., 1993', 'japan_quakes': 'Table of earthquakes around Japan from NOAA NGDC database', 'mars_shape': 'Table of topographic signature of the hemispheric dichotomy of  Mars from Smith and Zuber (1996)', 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa', 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis in Geology', 'ocean_ridge_points': 'Table of ocean ridge points for the entire world', 'rock_compositions': 'Table of rock sample compositions', 'usgs_quakes': 'Table of global earthquakes from the USGS'}
```
Then I try to use `pprint(list_sample_data())` instead. It returns:
```
pprint(list_sample_data())
{'bathymetry': 'Table of ship bathymetric observations off Baja California',
 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes',
 'fractures': 'Table of hypothetical fracture lengths and azimuths',
 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from  '
             'Mueller et al., 1993',
 'japan_quakes': 'Table of earthquakes around Japan from NOAA NGDC database',
 'mars_shape': 'Table of topographic signature of the hemispheric dichotomy '
               'of  Mars from Smith and Zuber (1996)',
 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis '
                          'in Geology',
 'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
 'rock_compositions': 'Table of rock sample compositions',
 'usgs_quakes': 'Table of global earthquakes from the USGS'}
```
Now it looks better but pylint reports that some lines are longer than 79 characters. So I only have two options:

1. Use `pprint(list_sample_data(), indent=75)` to make sure that the maximum line length < 80:
```
{'bathymetry': 'Table of ship bathymetric observations off Baja '
               'California',
 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with '
                       'holes',
 'fractures': 'Table of hypothetical fracture lengths and azimuths',
 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from '
             'Müller et al. (1993)',
 'japan_quakes': 'Table of earthquakes around Japan from the NOAA NGDC '
                 'database',
 'mars_shape': 'Table of topographic signature of the hemispheric '
               'dichotomy of Mars from Smith and Zuber (1996)',
 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data '
                          'Analysis in Geology',
 'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
 'rock_compositions': 'Table of rock sample compositions',
 'usgs_quakes': 'Table of earthquakes from the USGS'}
```
2. Use `pprint(list_sample_data(), indent=120)` so that each entry is printed in a single line:
```
{'bathymetry': 'Table of ship bathymetric observations off Baja California',
 'earth_relief_holes': 'Regional 20 arc-minutes Earth relief grid with holes',
 'fractures': 'Table of hypothetical fracture lengths and azimuths',
 'hotspots': 'Table of locations, names, and symbol sizes of hotpots from Müller et al. (1993)',
 'japan_quakes': 'Table of earthquakes around Japan from the NOAA NGDC database',
 'mars_shape': 'Table of topographic signature of the hemispheric dichotomy of Mars from Smith and Zuber (1996)',
 'maunaloa_co2': 'Table of CO2 readings from Mauna Loa',
 'notre_dame_topography': 'Table 5.11 in Davis: Statistics and Data Analysis in Geology',
 'ocean_ridge_points': 'Table of ocean ridge points for the entire world',
 'rock_compositions': 'Table of rock sample compositions',
 'usgs_quakes': 'Table of earthquakes from the USGS'}
```
I choose option 2 because in my opinion it's more readable, but I have to add `# noqa: W505` and `# pylint: disable=line-too-long` to disable flakeheaven and pylint errors:

File: pygmt/datasets/samples.py
Comment (seisman) [inline pygmt/datasets/samples.py:321]:
```suggestion
        :class:`xarray.DataArray` for raster data.
```

File: pygmt/datasets/samples.py
Comment (seisman) [inline pygmt/datasets/samples.py:320]:
```suggestion
        Sample dataset loaded as a :class:`pandas.DataFrame` for tabular data or
```

**Review by michaelgrund — APPROVED** [review]
Looks fine!

**Pull request merged**
